pkg/meta/model/job_args.go

```diff
@@ -1295,3 +1308,358 @@ func (a *FlashbackClusterArgs) decodeV1(job *Job) error {
+func GetAddIndexArgs(job *Job) (*AddIndexArgs, error) {
+	return getOrDecodeArgs(&AddIndexArgs{}, job)
+}
+
+func (a *AddIndexArgs) getAddIndexArgs(job *Job) error {
+	uniques := make([]bool, 1)
+	indexNames := make([]pmodel.CIStr, 1)
+	indexPartSpecifications := make([][]*ast.IndexPartSpecification, 1)
+	indexOptions := make([]*ast.IndexOption, 1)
+	hiddenCols := make([][]*ColumnInfo, 1)
+	globals := make([]bool, 1)
+
+	if err := job.DecodeArgs(
+		&uniques, &indexNames, &indexPartSpecifications,
+		&indexOptions, &hiddenCols, &globals); err != nil {
+		if err = job.DecodeArgs(
+			&uniques[0], &indexNames[0], &indexPartSpecifications[0],
+			&indexOptions[0], &hiddenCols[0], &globals[0]); err != nil {
+			return errors.Trace(err)
+		}
+	}
+
+	for i, unique := range uniques {
+		a.IndexArgs = append(a.IndexArgs, &IndexArg{
+			Unique:                  unique,
+			IndexName:               indexNames[i],
+			IndexPartSpecifications: indexPartSpecifications[i],
+			IndexOption:             indexOptions[i],
+			HiddenCols:              hiddenCols[i],
+			Global:                  globals[i],
+		})
+	}
+	return nil
+}
+
+func (a *AddIndexArgs) getAddPrimaryKeyArgs(job *Job) error {
+	a.IsPK = true
+	a.IndexArgs = []*IndexArg{{}}
+	if err := job.DecodeArgs(
+		&a.IndexArgs[0].Unique, &a.IndexArgs[0].IndexName, &a.IndexArgs[0].IndexPartSpecifications,
+		&a.IndexArgs[0].IndexOption, &a.IndexArgs[0].SQLMode,
+		&a.IndexArgs[0].Warning, &a.IndexArgs[0].Global); err != nil {
+		return errors.Trace(err)
+	}
+	return nil
+}
+
+func (a *AddIndexArgs) getAddVectorIndexArgs(job *Job) error {
+	var (
+		indexName              pmodel.CIStr
+		indexPartSpecification *ast.IndexPartSpecification
+		indexOption            *ast.IndexOption
+		funcExpr               string
+	)
+
+	if err := job.DecodeArgs(
+		&indexName, &indexPartSpecification, &indexOption, &funcExpr); err != nil {
+		return errors.Trace(err)
+	}
+
+	a.IsVector = true
+	a.IndexArgs = []*IndexArg{{
+		IndexName:               indexName,
+		IndexPartSpecifications: []*ast.IndexPartSpecification{indexPartSpecification},
+		IndexOption:             indexOption,
+		FuncExpr:                funcExpr,
+	}}
+	return nil
+}
+
+func (a *AddIndexArgs) getFinishedArgsV1(*Job) []any {
+	n := len(a.IndexArgs)
+	addIndexIDs := make([]int64, n)
+	ifExists := make([]bool, n)
+	isGlobals := make([]bool, n)
+	for i, arg := range a.IndexArgs {
+		addIndexIDs[i] = arg.AddIndexID
+		ifExists[i] = arg.IfExist
+		isGlobals[i] = arg.Global
+	}
+	return []any{addIndexIDs, ifExists, a.PartitionIDs, isGlobals}
+}
+
+// GetFinishedAddIndexArgs gets the add index args after the job is finished.
+func GetFinishedAddIndexArgs(job *Job) (*AddIndexArgs, error) {
+	if job.Version == JobVersion1 {
+		addIndexIDs := make([]int64, 1)
+		ifExists := make([]bool, 1)
+		isGlobals := make([]bool, 1)
+		var partitionIDs []int64
+
+		if err := job.DecodeArgs(&addIndexIDs[0], &ifExists[0], &partitionIDs, &isGlobals[0]); err != nil {
```

Comment:
when will this happen? getFinishedArgsV1 always returns a slice

Reply:
Currently, vector index returns single value. Could we change it to slice, since it's added in 8.4?

https://github.com/pingcap/tidb/blob/ffcfb5e08a36387e774c5990f637e964440de085/pkg/ddl/executor.go#L4714

Reply:
we have changed vector index to use v2 directly, so it cannot reach here, if that's the only reason then we can remove it

Reply:
If it doesn't feel like a big change, should we add v1 support for vector index? Mainly, the small constraint of "do not have a vector index during the v8.4-> (>= v8.5) upgrade" is quite difficult to deal with. Documentation needs to be written, and users may not necessarily see the documentation, and may also raise an issue, whether the problem is handled, etc.

Reply:
😓